### PR TITLE
Extensions: WPJM - Use redux-form to manage form state

### PR DESCRIPTION
**Pull request by donnapep**
There are two major changes involved with this PR:

1. The introduction of [Redux Form](http://redux-form.com/6.8.0/) to manage form state.
2. The removal of the `WPJobManager` component and the addition of a `Settings` component as its replacement. `Settings` uses `React.cloneElement` to create child components passed to it from the router. It is also responsible for rendering the form elements.

The below steps are needed in order to properly configure `redux-form`:

1. Add the [form reducer](http://redux-form.com/6.8.0/docs/GettingStarted.md/#step-1-of-4-form-reducer).
2. [Wrap the form component](http://redux-form.com/6.8.0/docs/GettingStarted.md/#step-2-of-4-form-component) with `reduxForm()`.
3. [Add `<Field/>` components](http://redux-form.com/6.8.0/docs/GettingStarted.md/#step-3-of-4-form-field-components) to connect each input to the store.
4. [Initialize values](http://redux-form.com/6.8.0/examples/initializeFromState/) received from the API.

There are three possible ways to add `<Field/>` components:

1. Using a [component](http://redux-form.com/6.8.0/docs/api/Field.md/#1-a-component)
2. Using a [stateless function](http://redux-form.com/6.8.0/docs/api/Field.md/#2-a-stateless-function)
3. Using a [string](http://redux-form.com/6.8.0/docs/api/Field.md/#3-a-string-input-select-or-textarea-)

Since we want to render our own custom form components and preserve the styling and other customizations, option 3 can be dismissed outright.

Using a component looked promising, but resulted in an [unknown prop warning](https://facebook.github.io/react/warnings/unknown-prop.html). This is because `redux-form` attaches `input` and `meta` props to the component. The `<FormTextInput/>` component (for example) [blindly copies those props](https://github.com/Automattic/wp-calypso/blob/master/client/components/forms/form-text-input/index.jsx#L44) to the `<input/>`, thereby triggering the warning. [Omitting them from the list](https://github.com/Automattic/wp-calypso/blob/master/client/components/forms/form-text-input/index.jsx#L34) results in the breaking of `redux-form` functionality.

That leaves using a stateless function, which is the approach this PR uses. The parent `<Settings/>` component takes care of rendering and updating the form elements using the `redux-form` API.

## Testing

To test, ensure that the form elements are initialized by data from the API, and that the form elements respond to user input.

### CODE REVIEW

**Comment by ockham** [comment]
Wow, pretty impressive! Thanks for taking this on! I think I'll need some time to let this fully sink in. 😄 

Some ad-hoc notes:

* I'm a bit surprised Redux Form passes component names as a prop to  `Field`, as opposed to wrapping components in a `Field` HOC. This seems to impose a rather rigid interface on the actual ('dumb') form components, and requires adapter methods like you provided in `Settings`. (I might be wrong, but I think HOCs might offer a bit more flexibility. Oh well.)
* Part of that `Settings` HOC is (and should be) so generic it shouldn't live inside any WPJM specific directory, but rather in `client/components`. Personally, I'd also opt for (small) single-purpose adaptors, i.e. one dedicated component (and file) per form component type. Can we encapsulate the `update...` methods in those individual components and drop the `cloneElement`/`Children.map()` that way?

**Comment by ockham** [comment]
> * Part of that `Settings` HOC is (and should be) so generic it shouldn't live inside any WPJM specific directory, but rather in `client/components`. Personally, I'd also opt for (small) single-purpose adaptors, i.e. one dedicated component (and file) per form component type. Can we encapsulate the `update...` methods in those individual components and drop the `cloneElement`/`Children.map()` that way?

I think I missed a crucial detail -- I mean composing those invidual components from what are now those `render...` (and `update...`) methods, plus our own corresponding `Form...` components, _plus_  `Field`. Would that work?

**Comment by donnapep** [comment]
@ockham I've created the Redux form components in `client/components/redux-forms`. What do you think of the naming? It would still enable us to switch to some other Redux form library if we wanted.

I couldn't get rid of the `Children.map()` code due to needing to pass `initialValues` to child components, and I doubt we would want to have each child add their own connected prop.

**Comment by ockham** [comment]
> @ockham I've created the Redux form components in `client/components/redux-forms`(https://github.com/Automattic/wp-calypso/pull/15619/commits/f497f6386f87819348530ea3b098b92772199711 & https://github.com/Automattic/wp-calypso/pull/15619/commits/19055f58ca13b2af388f3ae08fa0d8c82de15adf). What do you think of the naming? It would still enable us to switch to some other Redux form library if we wanted.

I like it :slightly_smiling_face:

**Comment by ockham** [comment]
Is it possible that you need to run `make shrinkwrap` (and commit the changed `npm-shrinkwrap.json `? It doesn't seem to be part of this branch yet.

File: client/extensions/wp-job-manager/components/data/query-settings/index.jsx [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+	render() {
+		return null;
+	}
+}
+
+QuerySettings.propTypes = {
+	siteId: PropTypes.number,
+	fetchingSettings: PropTypes.bool,
+	fetchSettings: PropTypes.func,
+};
```
Comment (ockham) [inline client/extensions/wp-job-manager/components/data/query-settings/index.jsx]: Should be inlined as `static`.

File: client/extensions/wp-job-manager/components/settings/index.jsx [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+const connectComponent = connect(
+	( state ) => {
+		const siteId = getSelectedSiteId( state );
+
+		return {
+			initialValues: getSettings( state, siteId ),
+			site: getSelectedSite( state ),
+			siteId,
```
Comment (ockham) [inline client/extensions/wp-job-manager/components/settings/index.jsx]: Nit -- It's a bit ugly to need both `site` and `siteId` from within the same component. `site` is only needed for `Navigation`, isn't it? Should we directly `connect` `Navigation` to it instead?

**Comment by ockham** [comment]
Hmm, this doesn't seem to fetch settings from the network for me:

image

Maybe some quirk in the `data-layer`? :roll_eyes:

**Comment by ockham** [comment]
Oh, I see `WP_JOB_MANAGER_ERROR` in my Redux log...

**Comment by ockham** [comment]
Feeling stupid now -- do we even have a WPJM `settings` endpoint yet? Can't seem to remember, and cursory scanning of the GH repo seems to indicate otherwise.

**Comment by donnapep** [comment]
> Feeling stupid now -- do we even have a WPJM settings endpoint yet? Can't seem to remember, and cursory scanning of the GH repo seems to indicate otherwise.

Oops, my bad. 🙁 Forgot to mention the REST API code can be found in [this branch](https://github.com/Automattic/wp-job-manager/tree/features/rest-api).

**Pull request merged**